### ISSUE TITLE
[DocDB] [Xcluster] [Tablet Splitting] Data inconsistency between Source and Target

### ISSUE THREAD

**Issue by shishir2001-yb** [yugabyte/yugabyte-db#19728]
Jira Link: [DB-8549](https://yugabyte.atlassian.net/browse/DB-8549)
### Description

Tried on version **2.20.0.0-b74**

Number of rows on Target 1008
Number of rows on Source 1001

For an Xcluster + Transactions + Aggressive tablet Splitting test case some of the deletes from the transaction are not being replicated on the target universe.

Note:
- This issue only happens in non-colocated tables; when we perform the same operation on a colocated table, this problem doesn’t manifest. The root cause appears to be related to Aggressive Tablet Splitting. Removing Aggressive Tablet Splitting eliminates this issue.

- This behaviour is specifically tied to transactions in which both update and delete operations are executed on the same row. Notably, this behaviour doesn’t manifest in every transaction.

- In the same transaction, other operations such as 10 inserts, 10 updates, and 9 deletes are properly reflected on the target.Additionally, introducing a significant time gap between different transaction commits seems to mitigate this issue.

- Wait On conflict, Deadlock Detection and Read committed were also enabled.

Test details :
```
- Create source and target Xcluster with g-flags for aggressive tablet splitting, Wait on Conflict, Deadlock detection, Read Committed
- Create both colocated and non-colocated table
- Create all types of index
- Create a backup on source and restore it on target
- Setup replication on both source and target
- Start a thread which will execute flush and compaction on all the tables after every 10 mins.
- Start 4 threads and perform the following operation for n number of times(There maybe some conflicts in DMLs):
        a. Insert 10 rows in both colocated and non-colocated table
        b. Update 10 rows in both colocated and non-colocated table
        c. Select 10 rows(FOR (UPDATE, SHARE, NO KEY UPDATE, KEY SHARE, "")) in both colocated and non-colocated table
        d. Delete 10 rows in both colocated and non-colocated table
        e. Commit the transaction 
- Validate data on both source and target
```
G-flags
```
t_flags.update({
                "enable_wait_queues": "true",
                "enable_deadlock_detection": "true",
                "enable_automatic_tablet_splitting": "true",
                "ysql_num_shards_per_tserver": 1,
                "enable_stream_compression": "true",
                "stream_compression_algo": "1",
                "ysql_output_buffer_size": "16214400",
                "scheduled_full_compaction_frequency_hours": "12",
                "scheduled_full_compaction_jitter_factor_percentage": "100",
                "scheduled_full_compaction_check_interval_min": "5",
                "yb_enable_read_committed_isolation": "true",
                "consistent_restore": "true"
            })
Master flags :
                "enable_wait_queues": "true",
                "enable_deadlock_detection": "true",
                "yb_enable_read_committed_isolation": "true",
                "enable_automatic_tablet_splitting": "true",
                "tablet_split_high_phase_shard_count_per_node": 200,
                "tablet_split_low_phase_shard_count_per_node": 16,
                "enable_stream_compression": "true",
                "stream_compression_algo": "1",
                "ysql_output_buffer_size": "16214400",
                "consistent_restore": "true"
                # high_phase_size 2MB
                # "tablet_split_high_phase_size_threshold_bytes": 2097152,
                # # low_phase_size 100KB
                # "tablet_split_low_phase_size_threshold_bytes": 102400,

```
The following rows were **deleted** in different transactions but the deletion was not replicated
```['80407', '767', '41787', '83105', '41895', '24231', '22119']```

[Universe Logs ](https://github.com/yugabyte/yugabyte-db/files/13207098/2.20.0.0_testreadcommittedwocxclusterpitrcolocationtabletsplitting-aws-rf3_20231030_084440.zip)

### Issue Type

kind/bug

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-8549]: https://yugabyte.atlassian.net/browse/DB-8549?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by hulien22** [comment]
sst-dump
```
SubDocKey(DocKey(0xd0e2, [20000008], []), [HT{ physical: 1698881900198798 w: 8 }]) -> DEL
SubDocKey(DocKey(0xd0e2, [20000008], []), [SystemColumnId(0); HT{ physical: 1698881900198798 w: 18 }]) -> null
SubDocKey(DocKey(0xd0e2, [20000008], []), [ColumnId(11); HT{ physical: 1698881900198798 w: 19 }]) -> "yyyyy"
SubDocKey(DocKey(0xd0e2, [20000008], []), [ColumnId(11); HT{ physical: 1698881900198798 w: 3 }]) -> "u"
```

```
jhe=# select * from test where i = 20000008;
    i     |   v
----------+-------
 20000008 | yyyyy
(1 row)
```

write id is out of order, so we end up hiding the update and delete

**Comment by hari90** [comment]
We found an issue in all version of xcluster (from 2.14), in both transactional and non-transactional modes where we miss some deletes.
If the source universe and target universe have different tablet counts, and we have batches that perform multiple operations on the same row in one transaction, then the last few operations are lost.
There is a ton of optimizations, fixes in the way intents are handled in xcluster, so I will keep the explanation as simple as possible and focus on only the write to the RegularDB here.
In the simplest form the test has 5 tablets on source, 1 on target and does a transaction in which 10 rows are inserted, updated and deleted . And we loose 4 deletes on the target. Lost in the sense that it never shows up however long we wait.
What is happening is that these 10 rows are written to 2 source tablets (6 keys in one and 4 on the other), and it results in 2 APPLY transaction WAL OPs (one in each tablet).  When these get xcluster eplicated to a single consumer tablet they show up as 2 batch from each tablet so we end up getting.
Batch1: Insert, Update 6 keys
Batch2: Insert, Update 4 keys
Batch3: Delete 6 keys, Apply Tran
Batch4: Delete 4 keys, Apply Tran
When we process the ApplyTran from Batch3 we collect all ops for this tran we have so far (Batch1, 2, 3) and write them to RocksDB in one batch with write_id 0-26.
Then when we process Batch4 we process the remaining ops (Batch4) and write them to RocksDB in one batch with write_id 0-4. If the 4 keys that we tried to delete here had used write-id >4 in the first apply then our delete op is ignored since it has a lower write_id!

In order to validate out theory we even tried using a hard coded write_id 0 each time, and the test passed (for this workload).
